src/mesh/amr_loadbalance.cpp

```diff
@@ -449,383 +698,205 @@ void Mesh::RedistributeAndRefineMeshBlocks(ParameterInput *pin, ApplicationInput
-              PARTHENON_MPI_CHECK(
-                  MPI_Test(&(req_recv[rb_idx]), &test, MPI_STATUS_IGNORE));
-              if (static_cast<bool>(test)) {
-                FinishRecvFineToCoarseAMR(pb.get(), recvbuf[rb_idx], loclist[on + l]);
-                received[rb_idx] = true;
+            auto pob = pb;
+            if (ranklist[on + l] == Globals::my_rank) pob = old_block_list[on + l - onbs];
+            LogicalLocation &oloc = loclist[on + l];
+            for (auto &var : pb->vars_cc_) {
+              if (!finished[idx]) {
+                auto var_in = pob->meshblock_data.Get()->GetVarPtr(var->label());
+                finished[idx] =
+                    TryRecvFineToCoarse(n - nbs, ranklist[on + l], oloc, var_in.get(),
+                                        var.get(), pb.get(), this);
               }
+              all_received = finished[idx++] && all_received;
             }
-            rb_idx++;
           }
-        } else { // c2f
-          if (ranklist[on] == Globals::my_rank) continue;
-          if (!received[rb_idx]) {
-            PARTHENON_MPI_CHECK(MPI_Test(&(req_recv[rb_idx]), &test, MPI_STATUS_IGNORE));
-            if (static_cast<bool>(test)) {
-              FinishRecvCoarseToFineAMR(pb.get(), recvbuf[rb_idx]);
-              received[rb_idx] = true;
+        } else if (oloc.level < nloc.level) { // c2f
+          for (auto &var : pb->vars_cc_) {
+            if (!finished[idx]) {
+              auto pob = pb;
+              if (ranklist[on] == Globals::my_rank) pob = old_block_list[on - onbs];
+              auto var_in = pob->meshblock_data.Get()->GetVarPtr(var->label());
+              finished[idx] = TryRecvCoarseToFine(
+                  n - nbs, ranklist[on], nloc, var_in.get(), var.get(), pb.get(), this);
             }
+            all_received = finished[idx++] && all_received;
           }
-          rb_idx++;
         }
       }
       // rb_idx is a running index, so we repeat the loop until all vals are true
-    } while (!std::all_of(received.begin(), received.begin() + rb_idx,
-                          [](bool v) { return v; }));
-    Kokkos::fence();
+    } while (!all_received && niter < 1e7);
+    if (!all_received) PARTHENON_FAIL("AMR Receive failed");
+  }
+  // Fence here to be careful that all communication is finished before moving
+  // on to prolongation
+  Kokkos::fence();
+
+  // Prolongate blocks that had a coarse buffer filled (i.e. c2f blocks)
+  for (int nn = nbs; nn <= nbe; nn++) {
+    int on = newtoold[nn];
+    LogicalLocation &oloc = loclist[on];
+    LogicalLocation &nloc = newloc[nn];
+    auto pmb = FindMeshBlock(nn);
+    if (nloc.level > oloc.level) {
+      const IndexDomain interior = IndexDomain::interior;
+      IndexRange cib = pmb->c_cellbounds.GetBoundsI(interior);
+      IndexRange cjb = pmb->c_cellbounds.GetBoundsJ(interior);
+      IndexRange ckb = pmb->c_cellbounds.GetBoundsK(interior);
+      // Need to restrict this block before doing sends
```

Comment:
This comment is not up-to-date any more isn't it?

Reply:
Oh, yeah. That looks like something I copy and pasted from the restriction above.